stl/inc/utility

```diff
@@ -647,6 +647,120 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+inline constexpr bool _Is_standard_integer = is_integral_v<_Ty>
+    && !_Is_any_of_v<remove_cv_t<_Ty>, bool, char,
+#ifdef _NATIVE_WCHAR_T_DEFINED
+        wchar_t,
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// FUNCTION TEMPLATE cmp_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_standard_integer<_Ty1> && _Is_standard_integer<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    if constexpr (is_signed_v<_Ty1> == is_signed_v<_Ty2>) {
+        return _Left == _Right;
+    } else if constexpr (is_signed_v<_Ty2>) {
+        return _Left == static_cast<make_unsigned_t<_Ty2>>(_Right) && _Right >= 0;
+    } else {
+        return static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right && _Left >= 0;
+    }
+}
+
+// FUNCTION TEMPLATE cmp_not_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !_STD cmp_equal(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE cmp_less
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_less(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_standard_integer<_Ty1> && _Is_standard_integer<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    if constexpr (is_signed_v<_Ty1> == is_signed_v<_Ty2>) {
+        return _Left < _Right;
+    } else if constexpr (is_signed_v<_Ty2>) {
+        return _Right > 0 && _Left < static_cast<make_unsigned_t<_Ty2>>(_Right);
+    } else {
+        return _Left < 0 || static_cast<make_unsigned_t<_Ty1>>(_Left) < _Right;
+    }
+}
+
+// FUNCTION TEMPLATE cmp_greater
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_greater(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return _STD cmp_less(_Right, _Left);
+}
+
+// FUNCTION TEMPLATE cmp_less_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_less_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !_STD cmp_less(_Right, _Left);
+}
+
+// FUNCTION TEMPLATE cmp_greater_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !_STD cmp_less(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE in_range
+template <class _Ty>
+_NODISCARD constexpr _Ty _Min_limit() noexcept { // same as (numeric_limits<_Ty>::min)(), less throughput cost

```

Comment:
Why not make it `_CONSTEVAL`?

Reply:
Thanks! I had simply forgotten about `_CONSTEVAL`'s existence. :joy_cat: Unfortunately, this triggered a compiler bug, which I've reduced and reported as Microsoft-internal VSO-1125263 "EDG rejects valid code using consteval":

```
C:\Temp>type kitty.cpp
#include <stdio.h>
#include <type_traits>
using namespace std;

#ifdef __cpp_consteval
template <typename T>
[[nodiscard]] consteval T MinLimit() noexcept {
    static_assert(is_integral_v<T>);
    if constexpr (is_signed_v<T>) {
        constexpr auto unsigned_max = static_cast<make_unsigned_t<T>>(-1);
        return static_cast<T>((unsigned_max >> 1) + 1);
    } else {
        return 0;
    }
}

template <typename T>
[[nodiscard]] consteval T MaxLimit() noexcept {
    static_assert(is_integral_v<T>);
    if constexpr (is_signed_v<T>) {
        constexpr auto unsigned_max = static_cast<make_unsigned_t<T>>(-1);
        return static_cast<T>(unsigned_max >> 1);
    } else {
        return static_cast<T>(-1);
    }
}

int main() {
    constexpr int min_int = MinLimit<int>();
    constexpr int max_int = MaxLimit<int>();
    printf("min_int: %d\n", min_int);
    printf("max_int: %d\n", max_int);
}
#else
    #error This test assumes that __cpp_consteval is defined.
#endif

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /c /BE kitty.cpp
kitty.cpp
"kitty.cpp", line 29: error: call to consteval function "MinLimit<T>() noexcept
          [with T=int]" did not produce a valid constant expression
      constexpr int min_int = MinLimit<int>();
                              ^
"kitty.cpp", line 29: note: cannot call non-constexpr function "MinLimit<T>()
          noexcept [with T=int]" (declared at line 7)
      constexpr int min_int = MinLimit<int>();
                              ^

"kitty.cpp", line 30: error: call to consteval function "MaxLimit<T>() noexcept
          [with T=int]" did not produce a valid constant expression
      constexpr int max_int = MaxLimit<int>();
                              ^
"kitty.cpp", line 30: note: cannot call non-constexpr function "MaxLimit<T>()
          noexcept [with T=int]" (declared at line 18)
      constexpr int max_int = MaxLimit<int>();
                              ^
```

Given that EDG is the only front-end defining `__cpp_consteval` right now, this effectively blocks us from using `_CONSTEVAL` here. (C1XX hasn't implemented it, and Clang 10's support is partial enough that they don't define the feature-test macro - although ironically, Clang 10 accepts this code if we don't query the macro.) I've recorded your excellent suggestion, so we can revisit this when one or more compilers actually accept `_CONSTEVAL` here.